[RTM] Symfony Session Handling

Toflar opened the pull request:
Requires the following changes in `contao/standard-edition`'s `config.yml`:

```diff
-    session: false
+    csrf_protection: ~
+    session:
+        handler_id: ~
```

(I did not make a separate PR because of the effort :-)).

I know it's mixing session handling with CSRF token handling but they sort of belong together and I had to make sure I could log in again, so I hacked it all together :-)

## ToDo's
* [x] Should we add an `ArrayAccess` wrapper for `$_SESSION`? Should writing and fetching from `$_SESSION['BE_DATA']` and `$_SESSION['FE_DATA']` still be supported? Ideas?
* [x] Should we add `public function getToken()` to `Template` because one should not use `REQUEST_TOKEN` anymore?
* [x] Tests
* [x] CS
* [x] Remove all internal usages of any PHP `session_*` function calls (e.g. in `User::authenticate()` - maybe @aschempp needs to have a look?)
* [x] Fix `Session::setData()` mapping
* [x] Check Request Token only if needed

Toflar commented:
I am now initializing the FE and BE attribute bags in the legacy Session class which is obviously wrong. The question is, should we provide those bags to the users (which means we add them via a generic listener by default)? Needs some discussion on Mumble. There's a lot of work to do here!

aschempp commented:
Looks good so far! I was not aware we can replace our request token with CSRF, but if that's possible I'd prefer to have that in a separate feature branch and merged as soon as it works (which appears to be fairly simple).

Toflar commented:
I won't be separating that. They belong together. If I implemented CSRF Handling first, I would need to configure and test native PHP session handling. If I do Session things first, I cannot log into the back end to see if session values are correctly stored. I'll keep it all together therefore.

Toflar commented:
Native session calls are removed or `FIXME`s are added where I'm not sure what to do. Also, the Contao frontend and backend session attribute bags are now registered via listener.

Toflar commented:
Our old Session class is now a simple mirror to the Symfony session :-) Now, I wanted to take on that `ArrayAccess` thing for `$_SESSION['BE_DATA']` and `$_SESSION['FE_DATA']` but I'm not sure if we should do that because if you did that, you did **not** use the API of Contao and you cannot expect this to work in a future version. Also I wonder if we should write that for `$_SESSION` because writing directly to it will still be supported and it will always be because it's a PHP superglobal. It's wrong but we cannot protect the users from doing it anyway so why should we add a component that "magically" maps them to the Symfony Session? I think it doesn't make sense. What do you think @contao/developers?

aschempp commented:
I think we should add the `ArrayAccess` mirror to `$SESSION['FE/BE_DATA']`. We know this is going to cause problems and there is no reason not to add it. It should obviously be deprecated in Contao 4 and removed in Contao 5 (maybe even add a deprecated notice).

Toflar commented:
> I think we should add the ArrayAccess mirror to $SESSION['FE/BE_DATA']. We know this is going to cause problems and there is no reason not to add it. It should obviously be deprecated in Contao 4 and removed in Contao 5 (maybe even add a deprecated notice).

@discordier and @leofeyer, what do you think?

a) Do we want the layer at all?
b) If yes, should it trigger an `E_USER_DEPRECATED`?

leofeyer commented:
What do we need the layer for?

Toflar commented:
Possible in Contao 3:
```php
$_SESSION['FE_DATA']['foo'] = 'bar';

echo \Session::getInstance()->get('foo');
```

This will not work in Contao 4 anymore. We could make it work by replacing `$_SESSION['FE_DATA']` and `$_SESSION['BE_DATA']` with a class implementing `ArrayAccess` and mapping it to the SF2 Session.

Toflar commented:
I have implemented everything (and more :-)) we've talked about:

* The `$_SESSION` keys `FE_DATA` and `BE_DATA` are now mirrored.
* I have added a FIXME for the token story
* User session is now invalidated
* I've replaced the `File::sendToBrowser()` by the SF `BinaryFileResponse` which is better suited than `StreamedResponse` and uses `stream_copy_to_stream`. This way I could get rid of the last native `session_*` call and I've added a `FIXME` there as the response must be thrown there using the upcoming `ResponseException`.

Removed `WIP` as this should be good to be merged now.

aschempp commented:
Maybe you should rebase to fix the merge conflicts?

Toflar commented:
That's because of 7b1d8995253edecdd71a11ee4d9eda1f861b009a, I can try :)

aschempp commented on contao/classes/RebuildIndex.php:
```diff
@@ -9,6 +9,8 @@
  */
 
 namespace Contao;
+use Symfony\Component\HttpKernel\KernelInterface;
```
missing empty line

aschempp commented on contao/controllers/BackendInstall.php:
```diff
@@ -9,6 +9,8 @@
  */
 
 namespace Contao;
+use Symfony\Component\HttpKernel\KernelInterface;
```
also missing empty line

aschempp commented on contao/controllers/BackendMain.php:
```diff
@@ -11,6 +11,8 @@
 namespace Contao;
 
 use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
```
here too …

aschempp commented on contao/library/Contao/File.php:
```diff
@@ -9,6 +9,8 @@
  */
 
 namespace Contao;
+use Symfony\Component\HttpFoundation\BinaryFileResponse;
```
another one

Toflar commented:
Stop commenting about CS, that's useless. Leo takes care of it anyway.

aschempp commented on contao/library/Contao/File.php:
```diff
@@ -681,32 +683,23 @@ public function resizeTo($width, $height, $mode='')
-		header('Content-Disposition: attachment; filename="' . ($filename ?: $this->basename) . '"');
-		header('Content-Length: ' . $this->filesize);
-		header('Cache-Control: must-revalidate, post-check=0, pre-check=0');
-		header('Pragma: public');
```
some of these headers are gone? Did you test if it still does open the "save as …" dialog?

Toflar replied:
Of course I tested! And the headers are either set by default via the BinaryFileResponse class or useless. I was pretty careful about that! It works like a charm for me :)

aschempp commented:
I wonder what's happening to CSRF if there is no security enabled? That's the case for the install tool for example.

aschempp commented on src/Session/Attribute/AttributeBagMirror.php:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @author Yanick Witschi <https://github.com/toflar>
+ */
+class AttributeBagMirror implements \ArrayAccess
```
I think this should be called `Adapter` instead of `Mirror` to follow the design pattern.

Toflar replied:
And I think it should be `Mirror` instead of `Adapter`.

aschempp commented:
> $feBag of type object<Symfony\Compon...SessionBagInterface> is not a sub-type of object<Symfony\Compon...ribute\AttributeBag>. It seems like you assume a concrete implementation of the interface Symfony\Component\HttpFo...ion\SessionBagInterface to be always present.

What does that mean?
https://scrutinizer-ci.com/g/contao/core-bundle/inspections/af5ac3b4-8bc7-4b59-bb7e-4e610ae17673/issues/files/src/EventListener/InitializeSystemListener.php?status=new&orderField=path&order=asc

Toflar commented:
> What does that mean?
> https://scrutinizer-ci.com/g/contao/core-bundle/inspections/af5ac3b4-8bc7-4b59-bb7e-4e610ae17673/issues/files/src/EventListener/InitializeSystemListener.php?status=new&orderField=path&order=asc

This is actually correct. `Session::getBag()` returns a `SessionBagInterface` instead of an `AttributeBagInterface` (which extends `SessionBagInterface`). Do I have to add a check or just `/* @var $feBag AttributeBagInterface */`to get rid of this?

Toflar commented:
> I wonder what's happening to CSRF if there is no security enabled? That's the case for the install tool for example.

I don't get the question but it does not seem like that's something I need to worry about.

aschempp commented:
> I wonder what's happening to CSRF if there is no security enabled? That's the case for the install tool for example.

I thought CSRF was related to security because the services are in that category. But according to your initial description `csrf_protection` is separately configured in the `config.yml` so that should be fine.

However, the unit tests coverage dropped significantly.

Toflar commented:
Yeah, no idea why. My code is 100% covered.

discordier commented on src/Resources/contao/library/Contao/File.php:
```diff
@@ -681,32 +683,23 @@ public function resizeTo($width, $height, $mode='')
 	 */
 	public function sendToBrowser($filename=null)
 	{
-		// Make sure no output buffer is active
```
Why is this in your PR?
Mind if I steal it into mine?

Toflar replied:
I can create a separate PR for it so @leofeyer can merge it first and we then both rebase?

discordier replied:
Might be also a proper solution.
Another thing, this PR is already "[RTM]" therefore if @leofeyer is willing to merge, I can rebase and add the exception.

leofeyer replied:
I'll rebase tonight or tomorrow.

leofeyer commented on src/EventListener/InitializeSystemListener.php:
```diff
@@ -151,6 +167,9 @@ private function boot($routeName, $basePath)
         $this->iniSet('error_log', $this->rootDir . '/system/logs/error.log');
 
+        // Support legacy Session access
+        $this->initializeLegacySessionAccess();
```
Why is the method call here and not in line 163, when `$this->startSession()` used to be?

Toflar replied:
Because I wanted this to be as early as possible. There's no specific reason - I would have added a comment if that was the case :)

leofeyer replied:
Then may I ask you to move it down to where it used to be? We should not mess with the initialization order too much, because we cannot foresee the effects.

Toflar replied:
I can, I just saw that it is now before `Config::preload();` which means that if somebody did something like
```php
$_SESSION['FE_DATA']['foo'] = 'bar';
```
in his `localconfig.php` it would still work. If I move it down, it won't work anymore.

aschempp replied:
It should probably be where the session was started previously?

aschempp replied:
I mean accessing the session in `localconfig.php` did not work before, because it was not started?

Toflar replied:
True, then I'll move it down.

Toflar replied:
Done in 63587f9d911a6bfb68e57a506ac3d4aa032108fa.

leofeyer commented on src/Resources/config/services.yml:
```diff
@@ -29,6 +31,14 @@ services:
+        arguments:
+            - "@session"
+        tags:
+            - { name: kernel.event_listener, event: kernel.request, method: registerContaoAttributeBags, priority: 64 }
```
Why did you set the priority to 64? We should add an explaining comment.

Toflar replied:
Added in 42686927457ea104cbd899bb6c660ffbad173732.

Toflar commented:
Rebased.

leofeyer commented:
Merged in d30310394fd209481ed6301f85d07e702300e444.
